suite/tests/security-common/selfmod.c

```diff
@@ -428,22 +555,61 @@ ADDRTAKEN_LABEL(last_byte_jmp_no_ilt:)
         jmp      last_byte_ret_zero     /* 1 byte opcode + 1 byte rel offset */
 last_byte_ret_one:
         mov      eax, HEX(1)
+#else
+        nop
+DECLARE_GLOBAL(last_byte_jmp_label)
+ADDRTAKEN_LABEL(last_byte_jmp_label:)
+        b last_byte_jmp
+last_byte_ret_zero:
+        mov x0, #0
+        ret
+last_byte_jmp:
+        b last_byte_ret_zero
+last_byte_ret_one:
+        mov x0, #1
+#endif
         ret
         END_FUNC(FUNCNAME)
 #undef FUNCNAME
 
 ALIGN_WITH_NOPS(16)
 
+#ifdef X86
 #define FUNCNAME make_last_byte_selfmod
         DECLARE_FUNC(FUNCNAME)
 GLOBAL_LABEL(FUNCNAME:)
+#ifdef X86
         lea      REG_XAX, SYMREF(last_byte_immed_plus_four - 4)
         mov      DWORD [REG_XAX], HEX(0)     /* selfmod write */
         mov      REG_XAX, HEX(0)             /* mov_imm to modify */
 ADDRTAKEN_LABEL(last_byte_immed_plus_four:)
+#else
+        adr x9, last_instruction - 4
+        ldr w10, [x9]
+        mov w11, #1
+        orr w10, w10, w11, LSL #5
+        str w10, [x9]
+        movz x9, #0 /* this will get updated by the code above */
+ADDRTAKEN_LABEL(last_instruction:)
+#endif
+        ret
+        END_FUNC(FUNCNAME)
+#undef FUNCNAME
+#else
+#define FUNCNAME make_code_page_selfmod
+        DECLARE_FUNC(FUNCNAME)
+GLOBAL_LABEL(FUNCNAME:)
+        adr x9, last_instruction - 4
+        ldr w10, [x9]
+        mov w11, #1
```

Comment:
Ditto.

Reply:
This function executes self-modifying code to create a sandboxed page. In this case the actual modification isn't significant.

Reply:
But DR won't actually do any "sandboxing" if it thinks the app has to issue explicit icache requests: DR would only do sandboxing when the hardware keeps the icache consistent (#5771).

Reply:
test_sandbox_last_byte() now calls clear_icache_if_required() which in the future will avoid calling tools_clear_icache() if it doesn't have to.